sdk/src/contexts/solana/relayer.ts

```diff
@@ -0,0 +1,186 @@
+import { Connection, PublicKey, PublicKeyInitData } from '@solana/web3.js';
+import { TokenBridgeRelayer } from './utils/types/tokenBridgeRelayer';
+import { createTokenBridgeRelayerProgramInterface } from './utils/tokenBridgeRelayer';
+import { BN, Program } from '@project-serum/anchor';
+import { ChainId } from 'types';
+import { NATIVE_MINT } from '@solana/spl-token';
+import {
+  RelayerFee,
+  deriveRelayerFeeAddress,
+  RegisteredToken,
+  deriveRegisteredTokenAddress,
+  RedeemerConfig,
+  deriveRedeemerConfigAddress,
+} from './utils/tokenBridgeRelayer/accounts';
+
+const SOL_DECIMALS = 9;
+const TEN = new BN(10);
+
+export interface SwapEvent {
+  recipient: string;
+  relayer: string;
+  token: string;
+  tokenAmount: string;
+  nativeAmount: string;
+}
+
+export class SolanaRelayer {
+  private program: Program<TokenBridgeRelayer>;
+
+  constructor(programId: PublicKeyInitData, private connection: Connection) {
+    this.program = createTokenBridgeRelayerProgramInterface(
+      programId,
+      connection,
+    );
+  }
+
+  async isAcceptedToken(mint: string): Promise<boolean> {
+    try {
+      const { isRegistered } = await this.getRegisteredToken(
+        new PublicKey(mint),
+      );
+      return isRegistered;
+    } catch (e) {
+      if (e instanceof Error && e.message?.includes('Account does not exist')) {
+        // the token is not registered
+        return false;
+      }
+      throw e;
+    }
+  }
+
+  async calculateRelayerFee(
+    targetChain: ChainId,
+    mint: PublicKey,
+    decimals: number,
+  ): Promise<bigint> {
+    const [{ fee }, { swapRate }, { relayerFeePrecision, swapRatePrecision }] =
+      await Promise.all([
+        this.getRelayerFee(targetChain),
+        this.getRegisteredToken(mint),
+        this.getRedeemerConfig(),
+      ]);
+    const relayerFee = TEN.pow(new BN(decimals))
+      .mul(fee)
+      .mul(new BN(swapRatePrecision))
+      .div(new BN(relayerFeePrecision).mul(swapRate));
+
+    return BigInt(relayerFee.toString());
+  }
+
+  async calculateMaxSwapAmountIn(
+    mint: PublicKey,
+    decimals: number,
+  ): Promise<bigint> {
+    const [
+      { swapRate, maxNativeSwapAmount },
+      { swapRate: solSwapRate },
+      { swapRatePrecision },
+    ] = await Promise.all([
+      this.getRegisteredToken(mint),
+      this.getRegisteredToken(NATIVE_MINT),
+      this.getRedeemerConfig(),
+    ]);
+    const swapRatePrecisionBN = new BN(swapRatePrecision);
+    const nativeSwapRate = this.calculateNativeSwapRate(
+      swapRatePrecisionBN,
+      solSwapRate,
+      swapRate,
+    );
+    const maxSwapAmountIn =
+      decimals > SOL_DECIMALS
```

Comment:
Does it need to be decouple in two cases? If `decimals < SOL_DECIMALS` the exponent `decimals - SOL_DECIMALS` will be negative, which is the same as dividing by `TEN.pow(SOL_DECIMALS - decimals)`. Perhaps there's something from the BN lib I'm missing

Reply:
Good catch, we can simplify this

Reply:
Actually, I changed this and it broke the gas slider. Here's a little example that shows that the two expressions aren't equivalent:

```
const TEN = new BN(10);
const SWAP_RATE_PRECISION = new BN(100_000_000);
const SOL_DECIMALS = 9;

function calculateSwapAmount(
  decimals: number,
  maxNativeSwapAmount: BN,
  nativeSwapRate: BN,
  firstCaseOnly: boolean = false
) {
  return decimals > SOL_DECIMALS || firstCaseOnly
    ? maxNativeSwapAmount
        .mul(nativeSwapRate)
        .mul(TEN.pow(new BN(decimals - SOL_DECIMALS)))
        .div(SWAP_RATE_PRECISION)
    : maxNativeSwapAmount
        .mul(nativeSwapRate)
        .div(TEN.pow(new BN(SOL_DECIMALS - decimals)).mul(SWAP_RATE_PRECISION));
}

  console.log(calculateSwapAmount(8, new BN(1000000), new BN(100000)));
  console.log(calculateSwapAmount(8, new BN(1000000), new BN(100000), true));

// prints
<BN: 64>
<BN: 2710>
```

Reply:
So after some digging I've discovered that bn.js simply doesn't support `pow()` using negative numbers, e.g.
```
> new BN(10).pow(new BN(-2)).toString()
'100'
```

Reply:
No choice but to leave it as is then